test/upgrade/installation/openshift.go

```diff
@@ -0,0 +1,92 @@
+package installation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/openshift-knative/serverless-operator/test"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+
+	configv1 "github.com/openshift/api/config/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func UpgradeOpenShift(ctx *test.Context) error {
+	clusterVersionName := "version"
+	clusterVersion, err := ctx.Clients.ConfigClient.ClusterVersions().Get(context.Background(), clusterVersionName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	var desiredImage string
+	if test.Flags.OpenShiftImage != "" {
+		desiredImage = test.Flags.OpenShiftImage
+		ctx.T.Logf("Desired OpenShift image: %s", desiredImage)
+	} else {
+		if len(clusterVersion.Status.AvailableUpdates) == 0 {
+			return errors.New("no OpenShift upgrades available")
+		}
+		desiredRelease := clusterVersion.Status.Desired
+		// Choose the highest version as the list can be unordered.
+		for _, update := range clusterVersion.Status.AvailableUpdates {
+			if update.Version > desiredRelease.Version {
+				desiredRelease = update
+			}
+		}
+		ctx.T.Logf("Desired OpenShift version: %s", desiredRelease.Version)
+		desiredImage = desiredRelease.Image
+	}
+	clusterVersion.Spec.DesiredUpdate = &configv1.Update{
+		Image: desiredImage,
+		Force: true,
+	}
+
+	if _, err = ctx.Clients.ConfigClient.ClusterVersions().Update(context.Background(),
+		clusterVersion, metav1.UpdateOptions{}); err != nil {
+		return err
+	}
+
+	ctx.T.Logf("Waiting for new cluster version to be ready...")
+	clusterVersion, err = WaitForClusterVersionState(ctx, clusterVersionName,
+		IsClusterVersionWithImageReady(desiredImage))
+	if err != nil {
+		return err
+	}
+	ctx.T.Logf("New cluster version is: %s", clusterVersion.Status.Desired.Version)
+
+	return nil
+}
+
+func WaitForClusterVersionState(ctx *test.Context, name string, inState func(s *configv1.ClusterVersion) bool) (*configv1.ClusterVersion, error) {
+	var lastState *configv1.ClusterVersion
+	var err error
+	waitErr := wait.PollImmediate(30*time.Second, 2*time.Hour, func() (bool, error) {
+		lastState, err = ctx.Clients.ConfigClient.ClusterVersions().Get(context.Background(), name, metav1.GetOptions{})
+		if apierrs.IsUnauthorized(err) {
```

Comment:
Is this really the only kind of errors during a cluster upgrade? (I'd expect even networking errors when api servers are restarted... maybe we should just log any kind of errors here and patiently wait... we can reconsider if we ever hit a case when a clear error is returned for the 2 hours...

Reply:
Yeah. Good point. That's the only error I saw during the cluster upgrade. We're only getting the ClusterVersion object. Maybe there are other errors in the cluster but we're only retrieving the ClusterVersion. What about this: keep this check as strict/narrow as possible and if we hit other errors we can reconsider the approach, after checking if the errors make sense. Rather than opening the doors completely. The errors will be visible in Jenkins because the build will fail. Anyway, either approach is good, IMO.